[Documentation][YouTube-Port] - YouTube Playlist & Videos Catalog in Port (#2)
paulappz/port-documentation: youtube_playlist.md

> # YouTube Playlist & Videos Catalog in Port

Reviewer: would not be needed.

---

## Step 3: Visualizing Data in Port

By leveraging Port's Dashboards, you can create custom dashboards to do the following : 

1. A dashboard for tracking playlist-level metrics like the number of videos.
2. Video-level insights, such as view count, position in playlist, and thumbnail displays.

<img src='/img/visualize.png' border='1px' />

### Dashboard setup

1. Go to your [software catalog](https://app.getport.io/organization/catalog).
2. Click on the `+ New` button in the left sidebar.
3. Select **New dashboard**.
4. Name the dashboard ( Visualise Youtube Playlist ), choose an icon if desired, and click `Create`.

This will create a new empty dashboard. Let's get ready-to-add widgets

### Adding widgets

<details>
<summary><b> Count of Videos in Playlist (click to expand)</b></summary>

1. Click `+ Widget` and select **Number Chart**.
2. Title: `Number of Videos`.(add the `Metric` icon).
3. Description: `Shows the number of videos on the playlist`(optional).
4. Select `Count entities` as Chart type.
5. Choose **Youtube Videos** as the **Blueprint**.
6. Select `count` for the **Function**.

<img src="/img/videocounts.png" border="1px" />

8. Click `Save`.

</details>

> 

Reviewer: You will have to rewrite this after the new implementation. do that and create a new pr so that the reviews on this are not too much.

# YouTube playlist & videos catalog in Port

This guide will help you set up an automated process to catalog YouTube playlist and video data into Port.

Using Port's GitHub action, you’ll fetch YouTube data and ingest it into Port for easy tracking and visualization.

  
## Prerequisites

> 1. [Create a Port account](https://app.getport.io) and set up API credentials.

Reviewer 2: ```suggestion
1. [Create a Port account](https://app.getport.io) and generate your API credentials.
```

2. [Obtain a YouTube Data API Key](https://console.cloud.google.com/apis/credentials).

> 3. [Set up GitHub secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets) in your repository for:

Reviewer 2: ```suggestion
3. [Set up GitHub secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets) in your repository for the following variables:
```

          }
        }
      },
      "videoOwnerChannelTitle": {
        "type": "string",
        "title": "Channel Title",
        "description": "The title of the channel that owns the video."
      },
      "videoOwnerChannelId": {
        "type": "string",
        "title": "Channel ID",
        "description": "The ID of the channel that owns the video."
      }
    },
    "required": [
      "videoDescription",
      "publishedAt",
      "duration",
      "link"
    ]
  },
  "mirrorProperties": {},
  "calculationProperties": {},
  "aggregationProperties": {},
  "relations": {
    "playlist": {
      "title": "Playlist",
      "many": false,
      "target": "youtube_playlist",
      "required": true
    }
  }
}

```

<center>
	<img  src='/img/data_model_blueprints.png'  border='1px'  />
</center>

</details>

## Setup GitHub actions workflow

1. Create `.github/workflows` in your repository.
2. Inside `.github/workflows`, create a YAML file (e.g., `youtube_port_workflow.yml`).

<details>

<summary>Example (click to expand)</summary>

```

<repository-root>/

      └── .github/

          └── workflows/

              └── <workflow-file>.yml

```

</details>

> 3. Define workflow in YAML.

Reviewer 2: ```suggestion
3. Define the workflow logic using the below GitHub YAML configuration.
```

<details>

<summary> Self-service action JSON: (click to expand)</summary>

```json
{
  "identifier": "create_youtube_catalog",
  "title": "Create Youtube Catalog",
  "icon": "Github",
  "description": "Self Service Action for Youtube Catalog Workflow",
  "trigger": {
    "type": "self-service",
    "operation": "CREATE",
    "userInputs": {
      "properties": {
        "playlist_id": {
          "icon": "DefaultProperty",
          "title": "Youtube Playlist Id",
          "type": "string"
        }
      },
      "required": [
        "playlist_id"
      ]
    },
    "blueprintIdentifier": "youtube_playlist"
  },
  "invocationMethod": {
    "type": "GITHUB",
    "org": "your-github-org",
    "repo": "your-github-repo",
    "workflow": "your-workflow-file.yml",
    "workflowInputs": {
      "{{ spreadValue() }}": "{{ .inputs }}",
      "port_context": {
        "entity": "{{.entity}}",
        "blueprint": "{{.action.blueprint}}",
        "runId": "{{.run.id}}",
        "trigger": "{{ .trigger }}"
      }
    },
    "reportWorkflowStatus": true
  },
  "requiredApproval": false
}
```

</details>

4. Click `Save`.

> 5. Click  `Create` button on just created action on the self service action page.

Reviewer 2: ```suggestion
5. Click  `Create` button on the just created action on the self service action page.
```

> 6. Enter the Playlist Id in the input feild. 

Reviewer 2: ```suggestion
6. Enter the `Playlist Id` in the input field.
```

> 7. Click  `Execute` button.

Reviewer 2: ```suggestion
7. Click  the `Execute` button to execute the action.
```

      "link"
    ]
  },
  "mirrorProperties": {},
  "calculationProperties": {},
  "aggregationProperties": {},
  "relations": {
    "playlist": {
      "title": "Playlist",
      "many": false,
      "target": "youtube_playlist",
      "required": true
    }
  }
}

```

<center>
	<img  src='/img/data_model_blueprints.png'  border='1px'  />
</center>

</details>

## Step 2: Setup GitHub Actions Workflow 

1. Create `.github/workflows` in your repository.
2. Inside `.github/workflows`, create a YAML file (e.g., `youtube_port_workflow.yml`).

<details>

<summary>Example (click to expand)</summary>

```

<repository-root>/

      └── .github/

          └── workflows/

              └── <workflow-file>.yml

```

</details>

3. Define Workflow in YAML

<details>
<summary>GitHub Workflow (click to expand)</summary>

```yaml showLineNumbers
name: Update YouTube Playlist and Video Entities in Port

on:
  workflow_dispatch:
    inputs:
      playlist_id:
        required: true
        description: "Youtube Playlist Id"
      port_context:
        required: false
        description: |
          Who triggered the action and general context (blueprint, run id, etc...)
        type: string 

>         

Reviewer 2: <img width="1504" alt="Screenshot 2024-11-19 at 8 37 01 AM" src="https://github.com/user-attachments/assets/fb6b6ecd-4f7c-4ef5-8a5f-cd022f9b2059">

I tried to execute your github workflow by using the self service action and it did not work. Please fix it.

this problem is coming from the self service action. You can fix it by adding the     `"{{ spreadValue() }}": "{{ .inputs }}",` to the workflowInputs as I've done below.

```
  "invocationMethod": {
    "type": "GITHUB",
    "org": "your org name",
    "repo": "your repo name",
    "workflow": "ingest-paul-youtube.yml",
    "workflowInputs": {
      "{{ spreadValue() }}": "{{ .inputs }}",
      "port_context": {
        "entity": "{{.entity}}",
        "blueprint": "{{.action.blueprint}}",
        "runId": "{{.run.id}}",
        "trigger": "{{ .trigger }}"
      }
    },
    "reportWorkflowStatus": true
  }
```

        with:
          clientId: ${{ secrets.PORT_CLIENT_ID }}
          clientSecret: ${{ secrets.PORT_CLIENT_SECRET }}
          baseUrl: https://api.getport.io
          operation: PATCH_RUN
          status: ${{ steps.bulk_create_update.outcome == 'success' && 'SUCCESS' || 'FAILURE' }}
          runId: ${{fromJson(inputs.port_context).runId}}
          logMessage: ${{ steps.bulk_create_update.outcome == 'success' && 'YouTube Data created/Updated Successfully' || 'Error in YouTube Data creation/update' }}  
```

</details>

4. Add and push the workflow file to your repository.
5. Go to **Actions** tab in GitHub to see workflow execution.
6. Go to **Settings** > **Secrets** to add Secrets used in the workflow.

  
<details>

<summary>Here’s an example of what you would see on Port calatog when Playlist and Video data has been injected. (click to expand)</summary>

<center>
<img  src='/img/playlist_catalog.png'  border='1px'  />
</center>
<center>
<img  src='/img/playlist_details.png'  border='1px'  />
</center>
<center>
<img  src='/img/videos_catalog.png'  border='1px'  />
</center>
<center>
<img  src='/img/videos_details.png'  border='1px'  />

> </center>

Reviewer 2: Don't you want to show this after you have walked them through how to configure the Port self service action. and after they trigger the action, and if everything is successful, you can tell them that is is an example of what they can see if the operation? At the moment, it has come too early in this guide

</details>

## Step 3: Create Youtube Catalog Blueprint using JSON definition.

<details>

<summary>Blueprint for `Youtube Catalog` Action (click to expand)</summary>

```yaml showLineNumbers
{
  "identifier": "youtubecatalog",
  "title": "YouTubeCatalogAutomation",
  "icon": "Github",
  "schema": {
    "properties": {
      "service_name": {
        "icon": "Github",
        "title": "Service Name",
        "type": "string",
        "description": "All Uppercase"
      }
    },
    "required": [
      "service_name"
    ]
  },
  "mirrorProperties": {},
  "calculationProperties": {},
  "aggregationProperties": {},
  "relations": {}
}

```

<center>
	<img  src='/img/catalog_blueprint.png'  border='1px'  />
</center>

> 

Reviewer 2: you need to delete this who section for blueprint for self service action. I've mentioned earlier that there is no such thing as blueprint for action.

2.  **Payload Sent**: A payload with metadata and inputs is sent to your infrastructure.
3.  **Job Triggered**: A job runs, and the user receives continuous progress updates.
4.  **Update Port**: The action's status, logs, and links are sent back to Port.

#### Step-by-Step Guide to create a Self-service Action for Youtube Playlist Workflow.

1. Navigate to the Self-service page and click the `+ New Action` button.
2. Click the `Edit JSON` button on the modal that appears.
4. Copy and paste the self-service action JSON (seen below) in the JSON section.
6. Click `Save`.

<details>

<summary> Self-service action JSON: (click to expand)</summary>

```json
{
  "identifier": "create_youtube_catalog",
  "title": "Create Youtube Catalog",
  "icon": "Github",
  "description": "Self Service Action for Youtube Catalog Workflow",
  "trigger": {
    "type": "self-service",
    "operation": "CREATE",
    "userInputs": {
      "properties": {
        "playlist_id": {
          "icon": "DefaultProperty",
          "title": "Youtube Playlist Id",
          "type": "string"
        }
      },
      "required": [
        "playlist_id"
      ]
    },

>     "blueprintIdentifier": "youtubecatalog"

Reviewer 2: `    "blueprintIdentifier": "youtubecatalog"
`you don't have any blueprint named youtubecatalog so this will not work. please update it

This guide will help you set up an automated process to catalog YouTube playlist and video data into Port.

Using Port's GitHub action, you’ll fetch YouTube data and ingest it into Port for easy tracking and visualization.

  
## Prerequisites

1. [Create a Port account](https://app.getport.io) and set up API credentials.
2. [Obtain a YouTube Data API Key](https://console.cloud.google.com/apis/credentials).
3. [Set up GitHub secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets) in your repository for:

-  `YOUTUBE_API_KEY`: Your YouTube API key.
-  `CLIENT_ID`: Your Port client ID.
-  `CLIENT_SECRET`: Your Port client secret.

  
## Step 1: Model Data in Port

> 

Reviewer: Add some small information here about data modeling and the reason this needs to be done here

### Data model setup
1. Navigate to your [Port Builder](https://app.getport.io/settings/data-model) page.
2. Click the `+ Blueprint` button to create a new blueprint.
3. Click the Edit JSON button on the Modal and

<details>
	<summary>Example of blueprint for `youtube_video` created (click to expand)	</summary>

<center>
	<img  src='/img/blueprint.png'  border='1px'  />
</center>

</details>

> 4. Add the Deployment blueprints in JSON format.

Reviewer: You can describe the need to add these blueprints, before you talk about adding them. just something brief.
and the Name should be like the title you give the blueprint
so this is Add the Youtube playlist blueprint

<details>

> <summary>Deployment blueprint for `youtube_playlist` (click to expand)</summary>

Reviewer: ```suggestion
<summary>Youtube playlist blueprint (click to expand)</summary>
```

            "type": "string",
            "title": "High Thumbnail",
            "description": "URL for the high-resolution thumbnail image."
          },
          "standard": {
            "type": "string",
            "title": "Standard Thumbnail",
            "description": "URL for the standard thumbnail image."
          }
        }
      },
      "localized": {
        "type": "object",
        "title": "Localized Information",
        "description": "Localized title and description for different regions.",
        "properties": {
          "title": {
            "type": "string",
            "title": "Localized Title",
            "description": "The localized title of the playlist."
          },
          "description": {
            "type": "string",
            "title": "Localized Description",
            "description": "The localized description of the playlist."
          }
        }
      }
    },
    "required": ["playlistTitle", "playlistDescription", "publishedAt", "channelId", "channelTitle"]    
    },    
    "mirrorProperties": {},
    "calculationProperties": {},
    "aggregationProperties": {},
    "relations": {}
  }

```

</details>

<details>

> <summary>Deployment blueprint for `youtube_video` (click to expand)</summary>

Reviewer: ```suggestion
<summary>Youtube video blueprint (click to expand)</summary>
```

            "type": "string",
            "title": "Standard Thumbnail",
            "description": "URL for the standard thumbnail image."
          },
          "maxres": {
            "type": "string",
            "title": "Max Resolution Thumbnail",
            "description": "URL for the maximum resolution thumbnail image."
          }
        }
      },
      "videoOwnerChannelTitle": {
        "type": "string",
        "title": "Channel Title",
        "description": "The title of the channel that owns the video."
      },
      "videoOwnerChannelId": {
        "type": "string",
        "title": "Channel ID",
        "description": "The ID of the channel that owns the video."
      }
    },
    "required": ["videoTitle", "videoDescription", "publishedAt", "duration", "link"]
  },
  "mirrorProperties": {},
  "calculationProperties": {},
  "aggregationProperties": {},
    "relations": {
      "playlist": {
        "title": "Playlist",
        "many": false,
        "target": "youtube_playlist",
        "required": true
      }
    }
  }

```

<center>
	<img  src='/img/data_model.png'  border='1px'  />
</center>

</details>

  
## Step 2: GitHub Workflow for Data Ingestion

> 

Reviewer: avoid blank headers without content

                               }')

            # Append each video JSON to the combined JSON array
            combined_json=$(echo $combined_json | jq --argjson video "$video_json" '. + [$video]')
          done

          # Save the combined JSON array to the environment variable for Port
          echo $combined_json > port_entities.json
          echo "entities=$(jq -c . port_entities.json)" >> $GITHUB_ENV

      - name: Bulk Create/Update YouTube Playlist and Video Entities in Port
        uses: port-labs/port-github-action@v1
        with:
          clientId: ${{ secrets.PORT_CLIENT_ID }}
          clientSecret: ${{ secrets.PORT_CLIENT_SECRET }}
          baseUrl: https://api.getport.io
          operation: BULK_UPSERT
          entities: ${{ env.entities }}

      - name: Inform completion of request to Create / Update Catalog in Port
        uses: port-labs/port-github-action@v1
        with:
          clientId: ${{ secrets.PORT_CLIENT_ID }}
          clientSecret: ${{ secrets.PORT_CLIENT_SECRET }}
          baseUrl: https://api.getport.io
          operation: PATCH_RUN
          status: "SUCCESS"
          runId: ${{fromJson(inputs.port_context).runId}}

>           logMessage: "Youtube Data created/Updated Successfully"

Reviewer 2: you just assumed that the operation above (bulk create or update playlist will be successful). you need to check if it was actually successful. you can use 

if: steps.<step-id>.outcome == 'success'

then you can set status to SUCCESS 

else, you can set status to FAILURE.

See an example here in the [Jira github workflow](https://docs.getport.io/guides/all/change-status-and-assignee-of-jira-ticket#github-workflow)

```

</details>

4. Add and push the workflow file to your repository.

5. Go to **Actions** tab in GitHub to see workflow execution.

6. Go to **Settings** > **Secrets** to add Secrets usedin the workflow.

  
<details>

<summary>Here’s an example of what you would see on Port calatog when Playlist and Video data has been injected. (click to expand)</summary>

<center>
<img  src='/img/playlist_catalog.png'  border='1px'  />
</center>
<center>
<img  src='/img/playlist_details.png'  border='1px'  />
</center>
<center>
<img  src='/img/videos_catalog.png'  border='1px'  />
</center>
<center>
<img  src='/img/videos_details.png'  border='1px'  />
</center>
</details>

  

> ## Step 3: Visualizing Data in Port

Reviewer 2: this section (step 3) is hidden under the step 2. I had to click on step 2 before I could see step 3. make sure to check if you have closed all your <details> tag so it displays well

By leveraging Port's Dashboards, you can create custom dashboards to do the following :
1. A dashboard for tracking playlist-level metrics like the number of videos.
2. Video-level insights, such as view count, position in playlist, and thumbnail displays.

<details>

<summary> Here’s an example of what you would see (click to expand)</summary>

<center>

> <img  src='/img/visualize.png'  border='1px'  />

Reviewer 2: I was expecting to see a playlist dashboard here not this page

</center>
 
</details>

### Dashboard setup

1. Go to your [software catalog](https://app.getport.io/organization/catalog).
2. Click on the `+ New` button in the left sidebar.
3. Select **New dashboard**.
4. Name the dashboard ( Visualise Youtube Playlist ), choose an icon if desired, and click `Create`.

This will create a new empty dashboard. Let's get ready-to-add widgets

### Adding widgets

> 

Reviewer 2: add a content explaining what you are about to do: creating a widget to display the number of counts in a video playlist

<details>
<summary><b> Count of Videos in Playlist (click to expand)</b></summary>

1. Click `+ Widget` and select **Number Chart**.
2. Title: `Number of Videos`.(add the `Metric` icon).
3. Description: `Shows the number of videos on the playlist`(optional).
4. Select `Count entities` as Chart type.
5. Choose **Youtube Videos** as the **Blueprint**.
6. Select `count` for the **Function**.

<center>
<img  src="/img/videocounts.png"  border="1px"  />
</center>
  
8. Click `Save`.

</details>

>   

Reviewer 2: can you create other widgets such as the average number of likes across all videos, the number of views a video has etc?

> This guide will help you set up an automated process to catalog YouTube playlist and video data into Port.

Author: The title was initially added at the beginning but was instructed to remove it.  I have added it back.

Using Port's GitHub action, you’ll fetch YouTube data and ingest it into Port for easy tracking and visualization.

  
## Prerequisites

1. [Create a Port account](https://app.getport.io) and set up API credentials.
2. [Obtain a YouTube Data API Key](https://console.cloud.google.com/apis/credentials).
3. [Set up GitHub secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets) in your repository for:

-  `YOUTUBE_API_KEY`: Your YouTube API key.
-  `CLIENT_ID`: Your Port client ID.
-  `CLIENT_SECRET`: Your Port client secret.

  
## Step 1: Model Data in Port

Data modeling in Port involves defining entities and their relationships to ensure consistent, structured data. 

This step is crucial as it allows your system to understand and manage data efficiently, ensuring that different entities (like playlists and videos) interact seamlessly within workflows.

### Data model setup
1. Navigate to your [Port Builder](https://app.getport.io/settings/data-model) page.
2. Click the `+ Blueprint` button to create a new blueprint.

> 3. Click the Edit JSON button on the Modal

Reviewer 2: ```suggestion
3. Click the `Edit JSON` button on the modal that appears
4. Copy and paste the blueprint schemas below into the blueprint editor

```

Author: Suggestion taken.

### Add Deployment blueprints.

Adding blueprints in Port helps define the structure and relationships of your data entities. 

> For example, by adding the YouTube playlist and Youtube video blueprints, you can ensure that the necessary properties for each blueprint (like title, description, and link) are captured and organized consistently within the system. This enables better data management and interaction with APIs.

Reviewer 2: delete this section. It makes the guide lengthy

      },
      "videoOwnerChannelId": {
        "type": "string",
        "title": "Channel ID",
        "description": "The ID of the channel that owns the video."
      }
    },
    "required": [
      "videoDescription",
      "publishedAt",
      "duration",
      "link"
    ]
  },
  "mirrorProperties": {},
  "calculationProperties": {},
  "aggregationProperties": {},
  "relations": {
    "playlist": {
      "title": "Playlist",
      "many": false,
      "target": "youtube_playlist",
      "required": true
    }
  }
}

```

<center>
	<img  src='/img/data_model_blueprints.png'  border='1px'  />
</center>

</details>

  
## Step 2: GitHub Workflow for Data Ingestion

A GitHub workflow for data ingestion automates the process of pulling data from external sources (like YouTube) and processing it for use within your application. 

By defining this workflow, you ensure that the data ingestion process is consistent, automated, and integrated into your development pipeline. This helps in efficiently handling data updates and maintaining data quality.

### Create a Youtube Catalog Blueprint using JSON definition.

<details>

> <summary>Deployment blueprint for `Youtube Catalog` Action (click to expand)</summary>

Reviewer 2: the Port self service action configuration should be under it's own section with an "##" header. At the moment, it is showing under the Github workflow for data ingestion.

Reviewer 2: Also, there is no such thing as "deployment" blueprint so rename that

2.  **Payload Sent**: A payload with metadata and inputs is sent to your infrastructure.

3.  **Job Triggered**: A job runs, and the user receives continuous progress updates.

4.  **Update Port**: The action's status, logs, and links are sent back to Port.

#### Step-by-Step Guide to create a Self-service Action for Youtube Playlist Workflow.

1. Navigate to the Self-service page and click the + New Action button.
2. Choose a name and icon for the action.
3. Define the inputs users need to fill out when executing the action.

  
**Action's Frontend:**

<details>
<summary>Self service action frontend (click to expand)</summary>

```json

  {
  "identifier": "create_youtube_catalog",
  "title": "Create YouTube Catalog",
  "icon": "Github",
  "description": "Self Service Action for YouTube Catalog Workflow",
  "trigger": {
    "type": "self-service",
    "operation": "CREATE",
    "userInputs": {
      "properties": {
        "service_name": {
          "icon": "DefaultProperty",
          "title": "Service Name",
          "type": "string"
        }
      },
      "required": [
        "service_name"
      ]
    }
  }

> }

Reviewer 2: we don't need to break this down into backend and frontend. it makes the guide very long and hard to read. so delete this and just present one JSON that has both the frontend, backend and approval

Author: Deleted.

```

</details>

**Define Action's Backend**

1. Define an **Invocation Method** of type **GITHUB** to define how the action will be executed.

2. Specify the payload to be sent to your handler.

<details>
<summary> Self service action backend (click to expand)</summary>

```json
{
  "invocationMethod": {
    "type": "GITHUB",
    "org": "your-github-org",
    "repo": "your-github-repo",
    "workflow": "your-workflow-file.yml",
    "workflowInputs": {
      "port_context": {
        "entity": "{{.entity}}",
        "blueprint": "{{.action.blueprint}}",
        "runId": "{{.run.id}}",
        "trigger": "{{ .trigger }}"
      }
    },
    "reportWorkflowStatus": true
  }
}
```

> </details>

Reviewer 2: delete this

**Set Guardrails (Optional)**

Manual Approvals: We instruct to set Manual approval while needed. but we will setup to false in this case.  

<details>
<summary>Manual approval congiration (click to expand)</summary>

```json
{
  "requiredApproval": false
}
```
</details>

> 

Reviewer 2: delete this

 
**Execute the Action**

Users can execute the action from the Port UI.

<details>

<summary>Self service action execution (click to expand)</summary>

```json
{
  "status": "SUCCESS",
  "logMessage": "YouTube Data created/Updated",
  "links": [
    {
      "name": "GitHub Workflow",
      "url": "https://github.com/your-github-org/your-github-repo/actions/runs/123456789"
    }
  ]
}
```

> </details>

Reviewer 2: we dont need to show this. please delete it

#### Action JSON Structure

> 

Reviewer 2: this is the action were are interested in, and that's what the user will create in Port. so all the above JSON breakdown is not necessary.

<summary><b>Count of Videos in Playlist (click to expand)</b></summary>

1. Click `+ Widget` and select **Number Chart**.
2. Title: `Number of Videos`. (add the `Metric` icon).
3. Description: `Shows the number of videos on the playlist` (optional).
4. Select `Count entities` for the **Chart type**.
5. Choose **YouTube Video** as the **Blueprint**.
6. Select `count` for the **Function**.  
7. Click `Save`.

    <center>
    <img src="/img/video_counts.png" border="1px" />
    </center>

</details>

#### Creating Other Widgets

You can also create other widgets to display additional data about your YouTube videos such as:

- **Average Number of Likes Across All Videos**

<details>
<summary><b>Average Number of Likes (click to expand)</b></summary>

1. Click `+ Widget` and select **Number Chart**.
2. Title: `Average Likes`. (add the `Metric` icon).
3. Description: `Shows the average number of likes across all videos` (optional).
4. Select `Agregate by property` for the **Chart type**.
5. Choose **YouTube Video** as the **Blueprint**.
6. Select `Like Count` for the **Property**.
6. Select `Average` for the **Function**.
7. Select `total` for the **Average of**.
8. Click `Save`.

>     <center>

Reviewer 2: <img width="953" alt="Screenshot 2024-11-18 at 5 11 07 PM" src="https://github.com/user-attachments/assets/bd3a84e4-8085-4520-b9c3-79344cfe922f">

this section doesn't look good to me. some are nested. make sure all are aligned
